A29-xds-tls-security.md

```diff
@@ -0,0 +1,340 @@
+
+##### Certificate Provider
+
+This is the actual plugin responsible for returning the requested certificates. The Certificate
+Provider is responsible for the following:
+
+- use the [Certificate Provider API](#certificate-provider-api) to fetch a signed cert from
+the service at periodic intervals configured for the plugin.
+
+- extract the root cert from the signed cert or use another API to retrieve the root cert
+from the service based on root cert expiry.
+
+- call back into the [Distributor](#distributor) to inform it of new certificate updates.
+
+##### Distributor
+
+A Distributor is a convenience class that "distributes" a single update from a
+[Certificate Provider](#certificate-provider) to multiple consumers. Also note that
+a Distributor is the only link between the Certificate Provider plugin code and its
+consumers which helps eliminate dependencies between the two pieces of code.
+
+A Distributor provides two interfaces:
+
+- an interface for the Certificate Provider to update certificate(s).
+
+- an interface for a consumer to register a watcher for watching certificate updates.
+
+For every incoming update from the Certificate Provider the Distributor will send out
+the same update to all its registered watchers. The Distributor can be considered as
+an extension of the Certificate Provider that allows a single Certificate Provider to
+support multiple watchers.
+
+#### gRPC Client Side Changes
+
+Let us see what changes are needed in the gRPC client architecture to use Certificate Provider
+plugins in the xDS use-case.
+
+The gRPC client xDS flow is described in
+[gRPC Client Architecture](./A27-xds-global-load-balancing.md#grpc-client-architecture).
+One of the steps in the flow involves receiving the Cluster resource in the
+[CDS response](./A27-xds-global-load-balancing.md#cds).
+When transport security (i.e. TLS/mTLS) is enabled, the response will include
+[UpstreamTlsContext](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L27)
+containing the TLS configuration needed by all the sub-channels in that cluster.
+The root-certificate information is contained in the validation context fields - either
+[validation_context_certificate_provider](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L190)
+or [combined_validation_context.validation_context_certificate_provider](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L145). For mTLS, it will also include local identity certificate
+information in [tls_certificate_certificate_provider](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L169).
+
+If a channel is using `XdsCredentials`, whenever its CDS balancer receives a
```

Comment:
I may be missing it, but I don't see any mention of how clients will do server authorization/hostname verification.

Reply:
This PR needs more work and your point will be addressed as part of that rework